File: nowcoder/二叉树的下一个结点.py
# encoding: utf-8

'''

@author: liangchi

@contact: bnu_llc@163.com

@software: pycharm

@file: 二叉树的下一个结点.py

@time: 2018/5/30 下午4:56

'''


def GetNext( pNode):
    # write code here
    if pNode.right is not None:
        pNode = pNode.right
        while pNode.left:
            pNode = pNode.left
        return pNode
    else:
        if pNode.next is None:
            return None
        if pNode.next.left is pNode:
            return pNode.next
        else:
            while pNode.next is not None and pNode.next.left != pNode:
                pNode = pNode.next
            if pNode.next is None:
                return None
            else:
                return pNode.next

if __name__ == '__main__':
    pass

File: sword_c++/二叉树的下一个结点.cpp
//
// Created by 李先生 on 2018/9/4.
//
#include <iostream>
#include <vector>
using namespace std;

struct TreeLinkNode {
    int val;
    struct TreeLinkNode *left;
    struct TreeLinkNode *right;
    struct TreeLinkNode *next;
    TreeLinkNode(int x) :val(x), left(NULL), right(NULL), next(NULL) {}
};

class Solution {
public:
    TreeLinkNode* GetNext(TreeLinkNode* pNode)
    {
        if (pNode->right != NULL)
        {
            TreeLinkNode *temp = pNode->right;
            while (temp->left)
                temp = temp->left;
            return temp;
        }
        else
        {
            if (pNode-> next == NULL)
                return  NULL;
            while(pNode->next && pNode->next->left != pNode)
            {
                pNode = pNode->next;
            }
            return pNode->next;
        }
    }
};

int main()
{

    return 0;
}